### PR TITLE
ISPN-7070

### PR DESCRIPTION
**Pull request by wburns**
CacheNotifierImplInitialTransferDistTest.testIterationBeganAndSegmentNotComplete
random failures

* Changed test to wait for both key and segment completion
* Fixed issue where iterator could be complete before segment was

https://issues.jboss.org/browse/ISPN-7070

### CODE REVIEW

File: core/src/main/java/org/infinispan/notifications/cachelistener/DistributedQueueingSegmentListener.java [hunk 1]
```diff
@@ -77,12 +77,15 @@ public boolean handleEvent(EventWrapper<K, V, CacheEntryEvent<K, V>> wrapped, Li
 
    @Override
    public void transferComplete() {
-      justCompletedSegments.forEach(completeSegment);
-      completed.set(true);
-      notifiedKeys.clear();
+      // Complete any segments that are still open - means we just didn't receive the completion yet
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/notifications/cachelistener/DistributedQueueingSegmentListener.java]: I'm having some trouble parsing this comment, could you elaborate on why we `completeSegment()` may not have been called for some of the segments?

File: core/src/main/java/org/infinispan/stream/impl/ClusterStreamManagerImpl.java [hunk 2]
```diff
@@ -452,8 +452,8 @@ public void forgetOperation(Object id) {
       }
 
       @Override
-      public void subscribe(Subscriber<? super V> s, Consumer<? super Supplier<PrimitiveIterator.OfInt>> onLostSegments,
-            Consumer<? super Supplier<PrimitiveIterator.OfInt>> onSegmentsComplete) {
+      public void subscribe(Subscriber<? super V> s, Consumer<? super Supplier<PrimitiveIterator.OfInt>> onSegmentsComplete,
+            Consumer<? super Supplier<PrimitiveIterator.OfInt>> onLostSegments) {
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/stream/impl/ClusterStreamManagerImpl.java]: `RemoteIteratorPublisher` default subscribe method still uses the old parameter order (although neither supplier does anything).

File: core/src/main/java/org/infinispan/stream/impl/DistributedCacheStream.java [hunk 3]
```diff
@@ -720,41 +718,36 @@ public void valueAdded(Object key) {
...
          // added for a response before the segments are completed.
          // The valueIterated method can be invoked at any point however.
          // See ClusterStreamManagerImpl$ClusterStreamSubscription.sendRequest where the added and segments are called into
-         AtomicBoolean shouldNotify = new AtomicBoolean(true);
+         ByRef<Supplier<PrimitiveIterator.OfInt>> segmentsToNotify = new ByRef<>(segments);
          Object key = currentKey.get();
          if (key != null) {
             pendingSegments.compute(key, (k, v) -> {
-               // This check is if iterator caught up to us
+               // The iterator has consumed all the keys, so there is no reason to wait and just notify of segment
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/stream/impl/DistributedCacheStream.java]: s/wait and/wait:/

File: core/src/main/java/org/infinispan/stream/impl/DistributedCacheStream.java [hunk 4]
```diff
@@ -720,41 +718,36 @@ public void valueAdded(Object key) {
...
-               shouldNotify.set(false);
-               if (v == null) {
-                  return segments;
-               } else {
-                  // This is ugly, but we shouldn't hit this often (need a response with no entries and completed
-                  // segments and iterator that hasn't caught up)
-                  return () -> Stream.of(segments, v).flatMapToInt(s -> toStream(s.get())).iterator();
-               }
+               // We save previous segments if there were any. This should normally be null since if we iterate
+               // over an entry then it will null this. So this is mostly for segments that have no entries
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/stream/impl/DistributedCacheStream.java]: If all the segments that will ever be set in `segmentsToNotify` are empty, then I'd make that clearer. Something like "valueIterated notifies the completion of non-empty segments, but we need to notify the completion of empty segments here".

**Comment by danberindei** [comment]
Unfortunately the test now fails in CI: http://ci.infinispan.org/job/Infinispan/job/PR-5598/1/

**Comment by wburns** [comment]
Oops I was just testing the remove one as I was able to get that one to fail about 2% of the time :) Guessing I just need to tweak the test method a bit more.

File: core/src/main/java/org/infinispan/stream/impl/ClusterStreamManagerImpl.java (hunk 2 above)
Comment (wburns) [reply core/src/main/java/org/infinispan/stream/impl/ClusterStreamManagerImpl.java]: Yeah this wasn't a bug, just variables in wrong spots :) I can fix the other.

File: core/src/main/java/org/infinispan/notifications/cachelistener/DistributedQueueingSegmentListener.java (hunk 1 above)
Comment (wburns) [reply core/src/main/java/org/infinispan/notifications/cachelistener/DistributedQueueingSegmentListener.java]: The reason it wouldn't be completed is that the completion is done on a different thread. I will see if I can tweak it to be done in a slightly different order to provide more clear ordering semantics.

File: core/src/main/java/org/infinispan/stream/impl/DistributedCacheStream.java (hunk 4 above)
Comment (wburns) [reply core/src/main/java/org/infinispan/stream/impl/DistributedCacheStream.java]: Sure

File: core/src/main/java/org/infinispan/stream/impl/DistributedCacheStream.java (hunk 3 above)
Comment (wburns) [reply core/src/main/java/org/infinispan/stream/impl/DistributedCacheStream.java]: Sure

**Comment by wburns** [comment]
Updated and tests should be fixed. I haven't gotten into the ordering of completion to be before iterator, but will keep looking. However this should be able to integrated as is even if it doesn't get changed.

**Pull request merged**